### ISSUE TITLE
Infer splits union types

### ISSUE THREAD

**Issue by yuyaryshev** [microsoft/TypeScript#47526]
# Bug Report

### 🔎 Search Terms

is:issue is:open infer union 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about 'infer'

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.0-dev.20220116#code/KYDwDg9gTgLgBAbwL4G4BQaD0m4EkB2AbgIYA2AlgCZzn4BmwUcd0cMAnmMHAK77kR8AZwzYAtBMlTpM2XPkLFS5SuVYcAEWB1a5GAPzrVxk6bPm1aWjEZ1iAY24AZYMToAeAGoA+RGjgBcELAMJ5kPMAAXHAAFISRngCUcAC8voQQVOhIGNa2DtwAwoIw7oUANHA+cKA2+JRCcC5uXr4I-oHBoeFRsfFJqemZlOiBcADmIYUAFuSklNExyWlwxfgwAPJQzR6F3tkYHFyrJVs77gCCvilwFzUgdQ0n6+60DEwVNPSMVb4A-M9Sp9qtFzld0GgjtwAELkcZbVJBGBQWjjOAAHzg+B4AFsAEY-TF4iAQUiufAY3j1bS0YDUTEaYg2CHiCxs9kczkSdR4YQwYjrchMgxGLli8US2QYeyCITwIQQHHAWHwqBhUgRaIqhE3AAM6BlfLghF10TWQJK7m1UEqcpR+HG3kq2PxjGuVMoNPwdLgxEaAvYLM0wBxsuRTOAjQA7tN2GxptxGFBWATSBAo3AccQANaRuCUch0d7AfCONAmgB0kxgMzmlCWVams3mDa66oiMQVSut7eAiRQY0HQ+H2DgABUAMpwKDASCwRoAcguUHGuJL8AgdDYnG4C5dBKgC5ojXwEHgfqEcPwxDxZLYEDgYGIUGISpsTE32+Oe+AhEYC4rI8ExnSoYFmRpyEaAl7GIHhgjgT88R4NFU3TNA4l1Rsa2betEiw2sW2SP0mlcDxrW8PC2x6TtFWVOEtl7ftB1HMcE0fZNbxDadI1JP8IK3JC0Ug5hyBAOkMB5AA5dM4DJGAF0aSZvRfGwglomooGTKBGhgB9gm4BwYB4MgvzzKNuGAgz6l9ewjJMsAOLJHFIjQQ05TgPFTQ8kkyQFRFK2rAjcIHYch1HQoNgAJUigBRQoxyHMQ4BizToEiGIACYADZKgAdmSScMoAZgyjLojHHc4AXc0rXom0kXtR0j2E09zyES9xmvTj71MqriVJckANcsMPIARmifrfIpG4AqbOsG0CnClhCuBR1wSSIuiuKxwAQkCJKUq09KMtyvKConYrSvKyqF3Ofc3Upc47VRXxMXOPhPR0b1KFekiWjsUhgl+85kQiYHSPcRkbG8ZqTzPX12qvG87103qF0mwaK3HG67txA8YePLF4YvJHutRqE+p8zGOlC2m6fp0LRxigANAAFbaYo0QcDtSqBjtOuB8vHC6SrK7Hv2tWGibajqupRh8KfRqmBQA8Xdw+r06Sl1qEdl5HuHJm6MZVisaYZ82LbCnBLYCXAKwNyrrSqgHSF9J9YCPFgmC7bg+GzU8owpGc-UEW1pggHh5g85wIfI2JyHtrHcddJhMWeh1KQ1r6fSJZWKQZCMKIwIA)

### 💻 Code

```ts
export {};

// Invalid infer for type unions

//----------------------------------------------------------------------------------------------------------------------
// Definition
//----------------------------------------------------------------------------------------------------------------------

interface Leaf<V> {
    setValue: (v:V) => void;
}

interface Cont<C, V> extends Leaf<V> {
    setValue: (v:V) => void;
    getChild: () => ContOrLeaf<C>;
}

type ContOrLeaf<A> = A extends Cont<infer C, infer V> ? Cont<C, V> : Leaf<A>;

type BigOr = string | number | boolean | undefined | Date;

//----------------------------------------------------------------------------------------------------------------------
// Instantiation
//----------------------------------------------------------------------------------------------------------------------

const someBigOrValue: BigOr = 0;
const v0: Cont<Cont<BigOr, string>, number> = undefined as any;

// Demonstrates why the error below makes difference
v0.getChild().getChild().setValue(someBigOrValue);                      // TS reports 'Argument of type 'number' is not assignable to parameter of type 'never'.' here, this is because of bug below
(v0.getChild().getChild() as Leaf<BigOr>).setValue(someBigOrValue);     // The problem resolves if bug is fixed


// Now let's generate some errors to see actual types we here and actual problem:
const b0: boolean = v0.getChild();              // CORRECT      - Error:(26, 7) TS2322: Type 'Cont<BigOr, string>' is not assignable to type 'boolean'.
const b1: boolean = v0.getChild().getChild();   // INCORRECT!   - Error:(27, 7) TS2322: Type 'Leaf<number> | Leaf<string> | Leaf<undefined> | Leaf<false> | Leaf<true> | Leaf<Date>' is not assignable to type 'boolean'. Type 'Leaf<number>' is not assignable to type 'boolean'.
                                                // EXPECTED     - Error:(27, 7) TS2322: Type 'BigOr' is not assignable to type 'boolean'. Type 'undefined' is not assignable to type 'boolean'.
                                                //                  I.e type BigOr 'fall apart' for some unknown reason, should be Leaf<BigOr> (i.e. Leaf<number | string | undefined | boolean | Date>)
```

### 🙁 Actual behavior

`v0.getChild().getChild()` gets type
`Leaf<number> | Leaf<string> | Leaf<undefined> | Leaf<false> | Leaf<true> | Leaf<Date>`
and because of this
`v0.getChild().getChild().setValue(someBigOrValue);`                      
Produces TS Error 'Argument of type 'number' is not assignable to parameter of type 'never'.' 

### 🙂 Expected behavior
`v0.getChild().getChild()` should have type
`Leaf<BigOr>` which also equals to `Leaf<number | string | undefined | boolean | Date>`

I expected this type because 
A extends Cont`<`**infer C**, infer V`>` ? Cont`<`**C**, V`>` : Leaf`<`A`>` applied to Cont`<`**BigOr**, string`>` should result in Cont`<`**BigOr**, string`>` and not what's above.

The code also demonstrates why 
`Leaf<BigOr>` and `Leaf<number> | Leaf<string> | Leaf<undefined> | Leaf<false> | Leaf<true> | Leaf<Date>`
is not the same.

In short: because the second one results in 'never' for all functions taking T argument.

**Comment by MartinJohns** [comment]
Look up distributed conditional types. This behaviour is intentional. The documentation provides an example for the syntax to opt out of this behaviour.

**Comment by yuyaryshev** [comment]
So the solution was just this:
```
type ContOrLeaf<A> = [A] extends [Cont<infer C, infer V>] ? Cont<C, V> : Leaf<A>;
```

[As mentioned here at the bottom of the page](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html)

That's great! Thanks a lot!

**Comment by yuyaryshev** [comment]
Issue closed.
